File: Algorithm/c_or_cpp/charpter1_linklist/level3/has_cycle.cpp
#include <stdlib.h>
#include <stdio.h>

#include <stdio.h>

struct ListNode
{
    int val;               // 代表数据
    struct ListNode *next; // 代表指针
};

struct ListNode *initLink(int a[], int size)
{
    int i;
    // 1、创建头指针
    struct ListNode *head = (struct ListNode *)malloc(sizeof(struct ListNode));
    head->val = a[0];
    head->next = NULL;
    // 头指针指向头结点
    struct ListNode *c = nullptr;
    c = head;

    struct ListNode *entrance = nullptr;
    // 3、每创建一个结点，都令其直接前驱结点的指针指向它
    for (i = 1; i < size; i++)
    {
        // 创建一个结点
        struct ListNode *newNode = (struct ListNode *)malloc(sizeof(struct ListNode));
        newNode->val = a[i];
        newNode->next = NULL;
        c->next = newNode;
        c = c->next;
        if (i == 1)
        {
            entrance = c;
        }
        if (i == size - 1)
        {
            c->next = entrance;
        }
    }

    return head;
}

void printList(struct ListNode *p)
{
    struct ListNode *temp = p; // temp指针用来遍历链表
    // 只要temp指向结点的next值不是NULL，就执行输出语句。
    while (temp)
    {
        // struct ListNode* f = temp;//准备释放链表中的结点
        printf("%d ", temp->val);
        temp = temp->next;
        // free(f);
    }
    printf("\n");
}

bool hasCycle(struct ListNode *head)
{
    if (head == NULL || head->next == NULL)
    {
        return false;
    }

    struct ListNode *fast = head;
    struct ListNode *slow = head;

    while (fast != NULL && fast->next != NULL)
    {
        fast = fast->next->next;
        slow = slow->next;

        if (fast == slow)
        {
            return true;
        }
    }

    return false;
}

struct ListNode *detectCycle(struct ListNode *head) {
    if (head == NULL) {
        return NULL;
    }
    
   struct ListNode* slow = head;
   struct ListNode* fast = head;
    
    while (fast != NULL) {
        slow = slow->next;
        
        if (fast->next != NULL) {
            fast = fast->next->next;
        } else {
            return NULL;
        }
        
        if (fast == slow) {
           struct ListNode* ptr = head;
            
            while (ptr != slow) {
                ptr = ptr->next;
                slow = slow->next;
            }
            
            return ptr;
        }
    }
    
    return NULL;
}

int main()
{
    struct ListNode *list1 = NULL;

    int a[] = {3, 2, 0, -4};

    list1 = initLink(a, 4);

    bool result = hasCycle(list1);

    return 0;
}

File: Algorithm/Leetcode/Code_Materials/python_source/charpter1_linklist/level3/has_cycle.py
# coding=utf-8
from python_source.list_node import ListNode


class HasCycle:
    # 确定环
    def hasCycle(self, head):
        seen = set()
        while head:
            if head in seen:
                return True
            seen.add(head)
            head = head.next
        return False

    # 确定环的入口
    def detectCycle(self, head):
        fast, slow = head, head
        while True:
            if not (fast and fast.next): return
            fast, slow = fast.next.next, slow.next
            if fast == slow: break
        fast = head
        while fast != slow:
            fast, slow = fast.next, slow.next
        return fast


if __name__ == '__main__':
    # 先构造题目要求的链表,简单构造即可
    a1 = ListNode(3)
    a2 = ListNode(2)
    a3 = ListNode(0)
    a4 = ListNode(-4)
    a1.next = a2
    a2.next = a3
    a3.next = a4
    # 构造环
    a4.next = a2

    # 判断是否存在环
    # hasCycle = HasCycle()
    # print hasCycle.hasCycle(a1)

    # 确定入口
    hasCycle = HasCycle()
    node = hasCycle.detectCycle(a1)
    print node.val
